File: api-cmssignature/csharp/src/validation/check/CryptoChecker.cs
using System;
using tr.gov.tubitak.uekae.esya.api.cmssignature.bundle;
using tr.gov.tubitak.uekae.esya.api.cmssignature.signature;
using tr.gov.tubitak.uekae.esya.api.common.util.bag;
using tr.gov.tubitak.uekae.esya.api.crypto.alg;
using tr.gov.tubitak.uekae.esya.api.crypto.parameters;
using tr.gov.tubitak.uekae.esya.api.crypto.provider.core.baseTypes;
using tr.gov.tubitak.uekae.esya.api.crypto.util;
/**
 * 
 * @author aslihan.kubilay
 *
 */
namespace tr.gov.tubitak.uekae.esya.api.cmssignature.validation.check
{
    public class CryptoChecker : BaseChecker
    {
        protected override bool _check(Signer aSigner, CheckerResult aCheckerResult)
        {
            aCheckerResult.setCheckerName(Msg.getMsg(Msg.SIGNATURE_CHECKER), typeof(CryptoChecker));//TODO bundle

            byte[] signedData = aSigner.getSignerInfo().getSignedAttributes().getBytes();

            if (signedData == null)
            {
                aCheckerResult.addMessage(new ValidationMessage(Msg.getMsg(Msg.SIGNED_ATTRIBUTES_ENCODE_ERROR))); //TODO bundle
                return false;
            }

            bool verified = false;
            try
            {
                PublicKey pubKey = new PublicKey(aSigner.getSignerCertificate().getSubjectPublicKeyInfo());
                DigestAlg digestAlg = DigestAlg.fromAlgorithmIdentifier(aSigner.getSignerInfo().getDigestAlgorithm());
                //SignatureAlg sigAlg = SignatureAlg.fromAlgorithmIdentifier(aSignerInfo.getSignatureAlgorithm());
                Pair<SignatureAlg, IAlgorithmParams> pair = SignatureAlg.fromAlgorithmIdentifier(aSigner.getSignerInfo().getSignatureAlgorithm());
                SignatureAlg sigAlg = pair.first(); 

                //Bazi imza uretiticileri signature alg kismina sadece asimetrik algoritmayi koyuyorlar.
                if (sigAlg == SignatureAlg.RSA_NONE && digestAlg != null)
                {
                    sigAlg = SignatureAlg.fromName(sigAlg.getAsymmetricAlg().getName() + "-with-" +
                            digestAlg.getName().Replace("-",""));
                }

                if (digestAlg == null)
                    throw new CMSSignatureException("Cant resolve digest algorithm from SignerInfo" + aSigner.getSignerInfo().getDigestAlgorithm());
                if (pair.first() == null)
                    throw new CMSSignatureException("Cant resolve signature algorithm from SignerInfo " + aSigner.getSignerInfo().getSignatureAlgorithm());

                //verified = SignUtil.verify(sigAlg, signedData, aSignerInfo.getSignature(), pubKey);
                verified = SignUtil.verify(sigAlg, pair.second(), signedData, aSigner.getSignerInfo().getSignature(), pubKey);

            }
            catch (Exception aEx)
            {
                aCheckerResult.addMessage(new ValidationMessage(Msg.getMsg(Msg.SIGNATURE_VERIFICATION_ERROR), aEx)); //TODO bundle
                return false;
            }

            if (!verified)
            {
                aCheckerResult.addMessage(new ValidationMessage(Msg.getMsg(Msg.SIGNATURE_VERIFICATION_UNSUCCESSFUL))); //TODO bundle
                return false;
            }

            aCheckerResult.setResultStatus(Types.CheckerResult_Status.SUCCESS);
            aCheckerResult.addMessage(new ValidationMessage(Msg.getMsg(Msg.SIGNATURE_VERIFICATION_SUCCESSFUL)));
            return true;

        }
    }
}

File: api-cmssignature/src/tr/gov/tubitak/uekae/esya/api/cmssignature/validation/check/CryptoChecker.java
package tr.gov.tubitak.uekae.esya.api.cmssignature.validation.check;

import static tr.gov.tubitak.uekae.esya.api.cmssignature.validation.Types.CheckerResult_Status.NOTFOUND;

import java.security.PublicKey;

import tr.gov.tubitak.uekae.esya.api.cmssignature.CMSSignatureException;
import tr.gov.tubitak.uekae.esya.api.cmssignature.bundle.CMSSignatureI18n;
import tr.gov.tubitak.uekae.esya.api.cmssignature.bundle.E_KEYS;
import tr.gov.tubitak.uekae.esya.api.cmssignature.signature.Signer;
import tr.gov.tubitak.uekae.esya.api.cmssignature.validation.Types.CheckerResult_Status;
import tr.gov.tubitak.uekae.esya.api.cmssignature.validation.ValidationMessage;
import tr.gov.tubitak.uekae.esya.api.common.util.bag.Pair;
import tr.gov.tubitak.uekae.esya.api.crypto.alg.DigestAlg;
import tr.gov.tubitak.uekae.esya.api.crypto.alg.SignatureAlg;
import tr.gov.tubitak.uekae.esya.api.crypto.params.AlgorithmParams;
import tr.gov.tubitak.uekae.esya.api.crypto.util.KeyUtil;
import tr.gov.tubitak.uekae.esya.api.crypto.util.SignUtil;

/**
 * 
 * @author aslihan.kubilay
 *
 */
public class CryptoChecker extends BaseChecker
{

	protected boolean _check(Signer aSigner, CheckerResult aCheckerResult)
	{
		aCheckerResult.setCheckerName(CMSSignatureI18n.getMsg(E_KEYS.SIGNATURE_CHECKER), CryptoChecker.class);

		byte[]signedData = aSigner.getSignerInfo().getSignedAttributes().getEncoded();

		if(signedData == null)
		{
			aCheckerResult.addMessage(new ValidationMessage(CMSSignatureI18n.getMsg(E_KEYS.SIGNED_ATTRIBUTES_ENCODE_ERROR))); 
			return false;
		}

		boolean verified = false;
		try 
		{
			PublicKey pubKey = KeyUtil.decodePublicKey(aSigner.getSignerCertificate().getSubjectPublicKeyInfo());
			DigestAlg digestAlg = DigestAlg.fromAlgorithmIdentifier(aSigner.getSignerInfo().getDigestAlgorithm());

			Pair<SignatureAlg, AlgorithmParams> pair = SignatureAlg.fromAlgorithmIdentifier(aSigner.getSignerInfo().getSignatureAlgorithm());
			
			SignatureAlg sigAlg = pair.first(); 
			//Bazı imza üretiticileri signature alg kısmına sadece asimetrik algoritmayı koyuyorlar.
			if (sigAlg == SignatureAlg.RSA_NONE && digestAlg != null)
			{
				sigAlg = SignatureAlg.fromName(sigAlg.getAsymmetricAlg().getName() + "-with-" + 
						digestAlg.getName().replace("-", ""));
			}
			
			if (digestAlg == null)
				throw new CMSSignatureException("Cant resolve digest algorithm from SignerInfo"+aSigner.getSignerInfo().getDigestAlgorithm());
			if (pair.first() == null)
				throw new CMSSignatureException("Cant resolve signature algorithm from SignerInfo "+aSigner.getSignerInfo().getSignatureAlgorithm());
			
			
			verified = SignUtil.verify(sigAlg,pair.second(), signedData, aSigner.getSignerInfo().getSignature(), pubKey);
		} 
		catch (Exception aEx) 
		{
			aCheckerResult.addMessage(new ValidationMessage(CMSSignatureI18n.getMsg(E_KEYS.SIGNATURE_VERIFICATION_ERROR),aEx));
			return false;
		}


		if(!verified)
		{
			aCheckerResult.addMessage(new ValidationMessage(CMSSignatureI18n.getMsg(E_KEYS.SIGNATURE_VERIFICATION_UNSUCCESSFUL)));
			return false;
		}

		aCheckerResult.setResultStatus(CheckerResult_Status.SUCCESS);
		aCheckerResult.addMessage(new ValidationMessage(CMSSignatureI18n.getMsg(E_KEYS.SIGNATURE_VERIFICATION_SUCCESSFUL)));
		return true;

	}
}
